libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp

```diff
@@ -91,3 +94,62 @@ void AP_BattMonitor_Backend::update_resistance_estimate()
     // update estimated voltage without sag
     _state.voltage_resting_estimate = _state.voltage + _state.current_amps * _state.resistance;
 }
+
+bool AP_BattMonitor_Backend::extended_arming_checks(bool report) const
+{
+    char failure_str[41]; // only 50 characters fit in statustext, reserve 9 for "Battery X"
+    float voltage;
+
+    switch (_params.failsafe_voltage_source()) {
+        case AP_BattMonitor_Params::BattMonitor_LowVoltageSource_Raw:
+        default:
+            voltage = _state.voltage;
+            break;
+        case AP_BattMonitor_Params::BattMonitor_LowVoltageSource_SagCompensated:
+            voltage = _mon.voltage_resting_estimate(_instance);
+            break;
+    }
+
+    if (_params._critical_voltage > 0 && voltage <= _params._critical_voltage) {
+        hal.util->snprintf(failure_str, ARRAY_SIZE(failure_str), "critical failsafe (%.2fV)", voltage);
+        goto failed;
+    }
+
+    if (_params._low_voltage > 0 && voltage <= _params._low_voltage) {
+        hal.util->snprintf(failure_str, ARRAY_SIZE(failure_str), "low failsafe (%.2fV)", voltage);
+        goto failed;
+    }
+
+    if (has_current()) {
+        float remaining_mah = _params._pack_capacity - _state.consumed_mah;
+        if (_params._critical_capacity > 0 && remaining_mah < _params._critical_capacity) {
+            hal.util->snprintf(failure_str, ARRAY_SIZE(failure_str), "critical failsafe (%.0f mAh)", remaining_mah);
+            goto failed;
+        }
+
+        if (_params._low_capacity > 0 && remaining_mah < _params._low_capacity) {
+            hal.util->snprintf(failure_str, ARRAY_SIZE(failure_str), "low failsafe (%.0f mAh)", remaining_mah);
+            goto failed;
+        }
+    }
+
+    if (!_state.healthy) {
+        hal.util->snprintf(failure_str, ARRAY_SIZE(failure_str), "is unhealthy");
+        goto failed;
+    }
+
+    if (_state.failsafe != AP_BattMonitor::BatteryFailsafe_None) {
+        hal.util->snprintf(failure_str, ARRAY_SIZE(failure_str), "is in a %s failsafe",
+                           _state.failsafe == AP_BattMonitor::BatteryFailsafe_Low ? "low" : "critical");
+        goto failed;
+    }
+
+    return true;
+
+failed:
+    if (report) {
+        gcs().send_text(MAV_SEVERITY_CRITICAL, "PreArm: Battery %d %s", _instance, failure_str);
```

Comment:
I would prefer the text is sent back to the caller, AP_Arming should be deciding what to do, not the library. That would mean that AP_Arming would be doing a loop running each instance checks, but I think that's fine.

Reply:
I'd rather AP_Arming didn't run the loop, because I'd like to be able to leave some checks at the AP_BattMonitor level. I can simply overwrite the return string everywhere in the internal calls and return the last one, but it requires AP_Arming to provide the buffer/buffer length which is rather annoying.